### DISCUSSION TITLE
Building and distributing congo-parser

### DISCUSSION THREAD

**Discussion by malakeel** [congo-cc/congo-parser-generator#38] [General]
Hello all, 

I am new to this project. I have done few small parsers with javacc long ago. I am interested in trying congo-parser. Looking at the source code, it I can see it a bit different compare to the current (or not too distant past) of javacc source organization. 
Mainly I noticed two things:
- Maven is used as a build system in javacc, while in congo-parser it is using ANT, which is fine.
So my question here, when (and if there's intention) we decide to distribute congo-parse-generator through maven repository, will this be done with Ivy ? especially for versioning.

- The newer javacc-8 seems to have separated generators, into their own modules, and made the parser-generation more modular. 
My second question, since the source code looks very different from javacc, will the core be somehow reduced, and introduce modularity, just like javacc-8 did ?

**Comment by stbischof** [comment]
Maven:

``` 
<build>
    <plugins>
      <plugin>
        <groupId>org.congocc</groupId>
        <artifactId>org.congocc.maven.plugin</artifactId>
        <version>2.0.0-RC1</version>
        <executions>
          <execution>
            <goals>
              <goal>ccc-generate</goal>
            </goals>
          </execution>
        </executions>
        <configuration>
          <grammarFile>${project.basedir}/src/main/ccc/Grammer.ccc</grammarFile>
          <lang>java</lang>
          <jdk>17</jdk>
          <quiet>false</quiet>
          <outputDir>${project.basedir}/src/gen/java</outputDir>
        </configuration>
      </plugin>
      <plugin>
        <groupId>org.codehaus.mojo</groupId>
        <artifactId>build-helper-maven-plugin</artifactId>
        <version>3.0.0</version>
        <executions>
          <execution>
            <id>add-source</id>
            <phase>generate-sources</phase>
            <goals>
              <goal>add-source</goal>
            </goals>
            <configuration>
              <sources>
                <source>src/gen/java</source>
              </sources>
            </configuration>
          </execution>
        </executions>
      </plugin>
    </plugins>
  </build>
```

**Reply by malakeel** [reply]
Thank you. This answers my question about having a maven plugin to start playing with congo-cc.

**Comment by stbischof** [comment]
What are your requiroments for versioning?

**Reply by malakeel** [reply]
@stbischof 
There's no requirements about versioning. I am just wondering if there's some sort of way to be able to tell which version I am using. I saw somewhere in the code, the parser generator checks to see if there's a new version. So I got a bit curious here.

**Comment by revusky** [comment]
First of all, CongoCC is a complete rewrite and has been for a couple of years.  The last part that was rewritten was the lexical generator part, and that [was complete in May of 2021](https://parsers.org/javacc21/lexical-code-rewritten/). So, one aspect of that is that trying to look at the Congo source code and compare it with the source of the legacy JavaCC project is not going to be very illuminating for the simple reason that the two codebases have, by now, diverged so radically.

In fact, looking at the code of the legacy JavaCC project is not going to be very illuminating. Let me expain something very basic about the situation. A parser generator is a tool that generates code -- for a parser, obviously. And the *only* way to implement such a tool in a maintainable way that properly *partitions* the problem -- *SoC*, *separation of concerns* -- is to use a template engine. This is simply a fact really. In the web application space, surely everybody understands that you can't build web pages using `out.println`. Well, you can, okay, but eventually the situation will devolve into an unmaintainable mess. You really do need some kind of templating system. In deep structure, generating parser code in Java or whatever language is not that different from generating web pages, and when you see a project like the legacy JavaCC that simply rejects the one tool (a template engine) that is needed to structure the code in a tractable way, then you know that this is basically hopeless.  So,it is actually readily obvious why that project has been spinning its wheels for a couple of decades. They went through 7 whole point release cycles, from a 1.x to a 7.x. (and they're talking about an 8.x) but none of that embodies a single new feature or idea.

Now, in terms of templating, the simplest way to understand what this is about is just to look at a few of the templates. For example, [here](https://github.com/congo-cc/congo-parser-generator/blob/main/src/templates/java/ParserProductions.java.ftl) is the FreeMarker template that generates the Java code that implements the parser productions of a grammar. The equivalent Python code is generated using [a different template](https://github.com/congo-cc/congo-parser-generator/blob/main/src/templates/python/parser_productions.inc.ftl). If you look at the Java code in the legacy JavaCC project that generates the equivalent code, it looks like [this](https://github.com/javacc/javacc/blob/master/src/main/java/org/javacc/parser/ParseEngine.java). I would say it is also instructive to eyeball the last [10 years of commit history of that file](https://github.com/javacc/javacc/commits/master/src/main/java/org/javacc/parser/ParseEngine.java). There are 28 commits on that page and that is over the course of the last 10 (!) years. But even that statistic understates how bleak the situation is. Most of those commits are nothingburger commits, like "fix trailing whitespace", things of that nature.

In fact, their code is so entangled that Sreeni, the project's self-appointed dictator for life, has stated that the core code of the project is a *no-go zone*. Nobody can go in there and do anything. Well, I guess this genius figured out that the one sure way of never making a mistake is to never do anything in the first place! So the core code of the project is a no-go zone -- it's Comanche Territory, it's the South Bronx, it's some part of the map that is blank with "here lurk dragons" written on it. Anyway, properly understood, Sreeni's "code generator" concept is really just a half-baked remedy for the problem (of his own creation) that he won't use a template engine. But if you actually look at his `CodeGenerator` stuff it's nearly as horrible as what it replaces. And, generally speaking, the guy's idea that the core code will be a "no go zone" and the project will make progress by various people working on their own "code generators" -- it's a pretty lame idea. Really it is. The interesting thing about that is he has been talking about this since 2015.  Go figure.

Now, I just noticed that you did [make a comment](https://github.com/javacc/javacc/discussions/227#discussioncomment-5844745) on their discussion forum. I guess you're right, insofar as there being no reason to think that moving from Maven to Gradle will remedy any of their problems. On the other hand, frankly, I don't get the feeling that you grasp just how disastrous the state of affairs there really is. (Or maybe you do...) But it really is unbelievable. In that vein, you might find this [blog post](https://parsers.org/rants-non-technical/a-bugs-life/) interesting. Obviously, a project in that state, it makes no difference whether they use Maven or Gradle! The whole concept is silly frankly, at least in this precise context.

Now, regarding this project, we actually do have an artifact on Maven Central. Stefan Bischof answered you separately about that. We have no plan to use Ivy, but really, even if we did use Ivy, it would be the same thing. You just download the jarfile from a server and use it. Well, one thing to understand about this project is that it is rather different in structure from a typical Java app. That is what [this blog post](https://parsers.org/javacc21/javacc-is-not-a-java-app/) is about. That page is a bit out of date (and needs some updating) but even so, it could be well worth your while to read it.

**Reply by malakeel** [reply]
@revusky 
Thank you for taking the time to comment here and there. I just shows a welcoming community. 
That being said, initially the ability to build context-aware language parser attracted my attention. I am not parser/compiler expert, but just to see what a context-aware language may look like. Diving into the code, again I saw few things that got me even more curious. Mainly the downloading of new jar thing, which I think it is something that better handled by build system. The second one is the tight coupling between generators and the rest of the code. I have noticed language specific code in the parser somewhere. Finally, the need to build twice is new to me. 

With regard to the other discussion on javacc project, I am not expecting this discussion to go anywhere really. May be I understand how such an important project got to this state, but this different story. Again, Maven Vs Gradle Vs Ant Vs Ivy is not a top that I enjoy discussing. I would work with any of them, and be happy about it. I was just trying to make a point that moving to another build, doesn't bring anything new. Instead, what java-cc needs is a separation between build phases. I agree with your point about template engine for javacc. At a minimum, this can be used to separate boot-strapping from actual build for the parser. Again, the idea of having javacc.jar that builds javacc.jar, then gets replaced by whatever, is a bit scary. If you have a bug somewhere, then you have multiple layers to dig into and find it. 

I will try to get my way around their source code, and at some point in the close future, I want to try congo-cc for a toy project. It looks very promising to me. I highly appreciate it. 

Thank you

**Reply by malakeel** [reply]
And by the way, I felt Sreeni and others are ready to work with others and accept some ideas. The main problem I see it is the confidence they have in their ability to review contributions. I think things may have gone side ways in the past, but getting back on track. The next issue would be finding a reasonable maintainer.

**Reply by revusky** [reply]
> And by the way, I felt Sreeni and others are ready to work with others and accept some ideas.

Quite honestly, I don't know where you got that *feeling*. The facts do speak for themselves. It has been 20 years and Sreeni was the project lead for the entire time. Well, in the beginning, if you look at the commit record, say 2004, 2005 or thereabouts, there was some bit of work from Sreeni to try to keep their Java language support up to date, but that petered out. Some Russian guy (a friend of Vladimir actually), his name is Ivan Ponomarev, donated a patch to get the Java language support up to JDK 7 level -- try with resources construct and a couple of other things -- and apparently it took them (Sreeni, I guess) something like a year to review the patch. Here is another interesting aspect of the Ponomarev patch: he only patched the JavaCC.jj file, not the JJTree.jjt file. So, currently, you can do a try-with-resources construct if you're using JavaCC, but *not* if you're using JJTree. It didn't occur to Sreeni that any such changes to the Java grammar need to be applied to both the JavaCC and the JJTree grammars.

(*That's pretty unserious, don't you think?*) Oh, and also, Sreeni has said that he sees no need for an `INCLUDE` directive.

> The main problem I see it is the confidence they have in their ability to review contributions. I think things may have gone side ways in the past, but getting back on track. 

Well, I'm sorry, I don't know how to put this very diplomatically. You're really just talking nonsense now. To think that people who have had 20 years to get their act together are suddenly going to "get back on track", this is really just delusional. Moreover, they can't get "back" on track, because they were never on track to start with! 

If you actually make a serious study of the history of that project, the time period in which it was actively developed is very short. I think it was first made available (as a binary distro, not open source) in mid-1996 and JJTree was added to the package in early 1997 and that's really the last significant feature added. Maybe there was some development into 1998 or so, but not much. When Sun open-sourced the code in mid-2003, the thing had been gathering dust for at least 5 years. I think Sreeni was somewhat involved in the original JavaCC project, but he was just a student intern. The project lead was Sriram Sankar, with whom I've had some correspondence actually, and he seems like a nice fellow, but he hasn't been involved with JavaCC for over 20 years.

> The next issue would be finding a reasonable maintainer.

Well, why would any self-respecting person want to take that on? It's one thing to take on the task of cleaning up some half-baked, buggy, obsolete codebase just for the heck of it, but to take that on when people (like Sreeni) are going to actively prevent you from doing anything????

But I do have to ask you a very basic question: Have you ever looked at the JavaCC code?

**Reply by revusky** [reply]
> Mainly the downloading of new jar thing, which I think it is something that better handled by build system

Well, that the tool checks for a newer version is not really any *fundamental feature*. It is just a convenience that is offered to the user. If you don't want the new version check, you just invoke it with the `-n` option. So the new version check is not even something that you are forced to use. If you have some other way of dealing with that, by all means. All that said, it's not really much of an issue. The `congocc.jar` has no dependencies besides the JDK itself. The code that the tool *generates* also has no dependencies. Now, I sense that you may balk at such a simple-minded disposition as this new-version check. But it works fine precisely because there are no dependencies to manage.  Or, in other words, that is all an up-to-date check needs to do, check for a newer .jar file. 

> Finally, the need to build twice is new to me.

There is no *need* to build twice. CongoCC (like JavaCC before it) is a tool that *self-bootstraps*. In other words, it is built using itself -- which means, in practice, a somewhat older version of itself. A full rebootstrap build/test is when you build the tool, run the various tests, and *then* you drop in the version you just built and check whether you can rebuild using that and retest.

Properly understood, the fact that you can take the version you just built and then rebuilt/retest *using that new build* is about the most demanding test of the system. However, it is not *necessary*. But the real point is that this is not some special characteristic of CongoCC. You can do a rebootstrap build/test with any system that is used to build itself. In principle anyway. So, that is worth understanding, but it is not a fundamental feature of this tool. Any self-bootstrapping tool -- i.e. that is built using itself -- in principle can be tested this way.

Earlier, I asked you this:

> But I do have to ask you a very basic question: Have you ever looked at the JavaCC code?

If you have not looked at the JavaCC code, I suggest you review [this file](https://github.com/javacc/javacc/blob/master/src/main/java/org/javacc/parser/NfaState.java). 

Also, this `CodeGenerator` concept that Sreeni is always on about, that is [here](https://github.com/javacc/javacc/blob/master/src/main/java/org/javacc/parser/CodeGenerator.java). Both [this file](https://github.com/javacc/javacc/blob/master/src/main/java/org/javacc/parser/LexGen.java), and [this one](https://github.com/javacc/javacc/blob/master/src/main/java/org/javacc/parser/ParseGen.java) subclass `CodeGenerator`. The first of those files invokes `genCode()`  317 times and the second one 509 times. I think it would be instructive to page through that. You could draw some conclusions about how *maintainable* an approach this `CodeGenerator` idea is. And you might ask yourself why anybody would want to be a *maintainer* of that codebase when there is a cleaned up version of all that, which is this project...

**Reply by revusky** [reply]
> The second one is the tight coupling between generators and the rest of the code.

I honestly don't know what you mean by this. You would need to point to specific places in the code and tell me why you perceive this as a *tight coupling*.

**Comment by revusky** [comment]
Well, the OP disappeared from the discussion quite a few months ago. So I guess I just close this finally.
